### ISSUE TITLE
ReferenceError: Property 'location' doesn't exist in React Native + Storybook

### ISSUE THREAD

**Issue by KrastanD** [kettanaito/msw#1777]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Node.js version

v18.18.0

### Reproduction repository

[Reproduction](https://github.com/KrastanD/rn-msw-example)

### Reproduction steps

Run the app in development. Notice that requests aren't being returned by msw. 

### Current behavior

I am trying to run msw for development in React Native. It works great for testing! For development I am using the setupServer from 'msw/native'. 
When msw is used I get the following error:
```
TypeError: undefined is not a function
    at createLazyCallback (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:597475:15)
    at InteractiveIsomorphicRequest (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:597331:66)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:589356:107)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:5194:17)
    at Promise (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:523942:17)
    at fetch (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:5121:25)
    at ?anon_0_ (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112592:33)
    at next (native)
    at asyncGeneratorStep (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4630:26)
    at _next (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4649:29)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4654:14)
    at Promise (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:523942:17)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4646:25)
    at apply (native)
    at _fetchData (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112596:32)
    at apply (native)
    at fetchData (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112588:32)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112598:16)
    at commitHookEffectListMount (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:72990:38)
    at commitPassiveMountOnFiber (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:74016:44)
    at commitPassiveMountEffects_complete (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:73988:40)
    at commitPassiveMountEffects_begin (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:73978:47)
    at commitPassiveMountEffects (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:73968:40)
    at flushPassiveEffectsImpl (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75162:34)
    at flushPassiveEffects (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75122:43)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75021:34)
    at workLoop (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77416:48)
    at flushWork (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77395:28)
    at performWorkUntilDeadline (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77601:48)
    at apply (native)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29730:26)
    at _callTimer (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29649:17)
    at _callReactNativeMicrotasksPass (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29679:17)
    at callReactNativeMicrotasks (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29842:44)
    at __callReactNativeMicrotasks (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2779:46)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2591:45)
    at __guard (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2763:15)
    at flushedQueue (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2590:21)
    at invokeCallbackAndReturnFlushedQueue (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2584:33)
```

### Expected behavior

I expect the request to be handled by the handlers but it never makes it there and just throws the error.

**Comment by KrastanD** [comment]
It was Storybook. Commenting out Storybook allows the requests to work as expected. I'll try to get them to work together and report back if I figure something out.

**Comment by kettanaito** [comment]
Hi, @KrastanD. Thanks for reporting this. Something odd is going on, we should look deeper into this. 

Also, do you get the same behavior using `msw@next` (#1464)? I highly recommend you try that version, it will be the official latest very soon.

**Comment by KrastanD** [comment]
Hi @kettanaito, 

I updated to `msw@next` and fixed the errors to get my tests to work but in development it fails in a similar way.

Commenting out Storybook doesn't completely solve the issue anymore though it gives a more descriptive warning: 
```
Possible Unhandled Promise Rejection (id: 0):
ReferenceError: Property 'location' doesn't exist
ReferenceError: Property 'location' doesn't exist
    at toAbsoluteUrl (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:525093:36)
    at methodCall (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:524744:45)
    at call (native)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:524648:35)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:5186:17)
    at tryCallTwo (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:61:9)
    at doResolve (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:216:25)
    at Promise (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:82:14)
    at fetch (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:5143:25)
    at ?anon_0_ (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112853:33)
    at next (native)
    at asyncGeneratorStep (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4652:26)
    at _next (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4671:29)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4676:14)
    at tryCallTwo (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:61:9)
    at doResolve (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:216:25)
    at Promise (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:82:14)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:4668:25)
    at apply (native)
    at _fetchData (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112857:32)
    at apply (native)
    at fetchData (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112849:32)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:112859:16)
    at commitHookEffectListMount (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:73221:38)
    at commitPassiveMountOnFiber (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:74247:44)
    at commitPassiveMountEffects_complete (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:74219:40)
    at commitPassiveMountEffects_begin (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:74209:47)
    at commitPassiveMountEffects (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:74199:40)
    at flushPassiveEffectsImpl (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75393:34)
    at flushPassiveEffects (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75353:43)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:75252:34)
    at workLoop (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77647:48)
    at flushWork (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77626:28)
    at performWorkUntilDeadline (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:77832:48)
    at apply (native)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29773:26)
    at _callTimer (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29692:17)
    at _callReactNativeMicrotasksPass (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29722:17)
    at callReactNativeMicrotasks (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:29885:44)
    at __callReactNativeMicrotasks (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2791:46)
    at anonymous (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2603:45)
    at __guard (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2775:15)
    at flushedQueue (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2602:21)
    at invokeCallbackAndReturnFlushedQueue (http://192.168.0.101:19000/index.bundle?platform=ios&dev=true&hot=false:2596:33)
```

When I uncomment out Storybook the warning upgrades to an error that just says:
```
Unhandled promise rejection [ReferenceError: Property 'location' doesn't exist]
```

I could be wrong but my guess is that this is referring to window.location (we don't have that in node/react native). Let me know what you think! Thanks!

**Comment by kettanaito** [comment]
@KrastanD, it would be nice to narrow this error to MSW specifically. We are doing our best to be very careful around `window` and other globals, checking them before accessing them. Perhaps someone from the Storybook team can confirm if they know something about this. cc @yannbf.

**Comment by KrastanD** [comment]
I'm using @storybook/react-native so maybe @dannyhw would be able to help more. I'll try to reproduce it in a repo for both stable and next versions of msw so it's easier for everyone to take a look and isolate the issue.

**Comment by dannyhw** [comment]
Once you have the repro I'll take a look. I know that other people have had success with msw + react native so it should work in theory.

**Comment by KrastanD** [comment]
I created the reproduction in this [repo](https://github.com/KrastanD/rn-msw-example). The readme explains it in more detail but at a high level, I have `main` which works fine in tests and dev and three branches `with-storybook`, `msw-2.0` and `msw-2.0-with-storybook` which all don't work in development but do work in tests.

**Comment by dannyhw** [comment]
I looked at your repro and I was wondering something, I saw that you use msw/node but I thought there was a 'msw/native' which is for react native?

**Comment by KrastanD** [comment]
In the `mocks` folder I have `testServer.js` which uses msw/node and `devServer.js` which uses msw/native. In `setupJest.js` I use the testServer and in `App.js` I use the devServer

**Comment by Splanis** [comment]
Did you find any solution? I face the same issue, I have the `msw-storybook-addon` which runs the `setupServer` when calling `initialize` and the requests are not being sent at all, I get this error.

**Comment by yannbf** [comment]
Hey @Splanis can you share a reproduction? I had trouble using @KrastanD's reproduction (`[GraphQL] Entity not authorized: AccountEntity` when running storybook).

Are you using Storybook for React native as well?

**Comment by Splanis** [comment]
> Hey @Splanis can you share a reproduction? I had trouble using @KrastanD's reproduction (`[GraphQL] Entity not authorized: AccountEntity` when running storybook).
> 


I don't have unfortunately. 
To be clear, when the setupServer is running, i get this error on ALL requests and I don't have any handlers. Is this intended or not?

I created a handler at storybook for a specific request but i doesn't seem to run, I get the same error.


> Are you using Storybook for React native as well?
>
yes

**Comment by KrastanD** [comment]
@yannbf I'm not sure what that error you are getting is since I am not using GraphQL or have any AccountEntity. The repro is as basic of a react-native project as I could make. Feel free to create an issue on the repo tho and I'd be happy to help figure it out.

**Comment by kettanaito** [comment]
A bit of context:

- The `ReferenceError: Property 'location' doesn't exist` error originates from [XMLHttpRequest interceptor](https://github.com/mswjs/interceptors/blob/e868e460142be6874538320f8264996166961fb3/src/interceptors/XMLHttpRequest/XMLHttpRequestController.ts#L598-L600) that depends on the `location` global object to be present. XHR is primarily designed for browser-like environments in Node.js that must also guarantee other browser globals implementation, like `location. 

But since we are in React Native, I presume global `location` doesn't exist, does it? MSW still uses `XMLHttpRequestInterceptor` for React Native because that's how RN constructs requests:

https://github.com/mswjs/msw/blob/15e3006cdbcf1d23312252960ee2920d2011d1bb/src/native/index.ts#L16

In fact, that's the only interceptor that will work in React Native. 

It looks like we cannot depend on `location` entirely in the XHR interceptor. In the context of React Native, relative URLs are impossible anyway (nothing to be relative to). I believe we can have a safer check for `location` in `toAbsoluteUrl` and return the input URL as-is if the `location` is not present.

**Comment by kettanaito** [comment]
The original error `undefined it not a function` is not a thing since 2.0. The new release of MSW doesn't ship the `createLazyCallback` function anymore.

**Comment by kettanaito** [comment]
## Released: v2.0.5 🎉

This has been released in v2.0.5!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.5)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.5)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Comment by KrastanD** [comment]
I am now facing #1775 which doesn't look like it will be solved soon/easily. If that's the case, react native users should use v1 which has the original `undefined is not a function` issue. @kettanaito would you release a patch version to v1 if I were to look into whatever is causing the issue in `createLazyCallback`?
